File: dong_image_viewer/core/vfs.py
from __future__ import annotations

from pathlib import Path

from .archive import ArchiveReader, to_vfs_items
from .models import ItemType, VfsItem, classify_path
from .security import ArchivePolicy
from .uri import file_uri, is_archive_uri, parse_archive_uri, parse_file_uri


class VirtualFileSystem:
    def __init__(self, policy: ArchivePolicy | None = None):
        self.archive_reader = ArchiveReader(policy)

    def open_collection(self, target: str | Path) -> list[VfsItem]:
        uri = str(target)
        if not uri.startswith("file://") and not uri.startswith("archive://"):
            path = Path(target)
            uri = file_uri(path)
        if is_archive_uri(uri):
            return self._list_archive_uri(uri)

        path = parse_file_uri(uri)
        if path.is_dir():
            return self._list_directory(path)
        item_type = classify_path(path)
        if item_type == ItemType.ARCHIVE:
            entries = self.archive_reader.list(str(path))
            return to_vfs_items(file_uri(path), entries)
        if item_type == ItemType.IMAGE:
            return self._list_directory(path.parent, selected=path)
        return []

    def read_bytes(self, uri: str) -> bytes:
        if is_archive_uri(uri):
            parsed = parse_archive_uri(uri)
            if is_archive_uri(parsed.archive_uri):
                parent_bytes = self.read_bytes(parsed.archive_uri)
                return self.archive_reader.read_member(parsed.archive_uri, parsed.member_path, parent_bytes)
            archive_path = parse_file_uri(parsed.archive_uri)
            return self.archive_reader.read_member(str(archive_path), parsed.member_path)

        path = parse_file_uri(uri)
        return path.read_bytes()

    def _list_directory(self, path: Path, selected: Path | None = None) -> list[VfsItem]:
        items: list[VfsItem] = []
        for child in sorted(path.iterdir(), key=lambda p: (not p.is_dir(), p.name.lower())):
            if child.is_dir():
                item_type = ItemType.DIRECTORY
            else:
                item_type = classify_path(child)
            if item_type == ItemType.OTHER:
                continue
            stat = child.stat()
            items.append(
                VfsItem(
                    uri=file_uri(child),
                    display_name=child.name,
                    item_type=item_type,
                    size=stat.st_size,
                    compressed_size=stat.st_size,
                    modified_at=stat.st_mtime,
                )
            )
        if selected and all(item.uri != file_uri(selected) for item in items):
            stat = selected.stat()
            items.append(
                VfsItem(
                    uri=file_uri(selected),
                    display_name=selected.name,
                    item_type=ItemType.IMAGE,
                    size=stat.st_size,
                    compressed_size=stat.st_size,
                    modified_at=stat.st_mtime,
                )
            )
        return items

    def _list_archive_uri(self, uri: str) -> list[VfsItem]:
        parsed = parse_archive_uri(uri)
        archive_bytes = self.read_bytes(uri)
        entries = self.archive_reader.list(parsed.member_path, archive_bytes)
        return to_vfs_items(uri, entries)

File: src/core/Vfs.cpp
#include "Vfs.h"

#include "Uri.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <algorithm>
#include <stdexcept>

namespace {
void ensureNestedDepthAllowed(const QString& uri, const ArchivePolicy& policy)
{
    if (archiveNestedDepth(uri) > policy.maxNestedDepth) {
        throw std::runtime_error("Archive nesting depth limit exceeded");
    }
}
}

Vfs::Vfs(ArchivePolicy policy)
    : policy_(policy)
    , zipReader_(policy)
{
}

std::vector<VfsItem> Vfs::openCollection(const QString& target) const
{
    QString uri = target;
    if (!uri.startsWith("file://") && !uri.startsWith("archive://")) {
        uri = fileUriFromPath(target);
    }
    if (isArchiveUri(uri)) {
        return listArchiveUri(uri);
    }

    const QString path = pathFromFileUri(uri);
    const QFileInfo info(path);
    if (info.isDir()) {
        return listDirectory(path);
    }

    const VfsItemType type = classifyPath(path);
    if (type == VfsItemType::Archive) {
        QFile file(path);
        if (!file.open(QIODevice::ReadOnly)) {
            throw std::runtime_error("Failed to open archive");
        }
        return zipReader_.list(fileUriFromPath(path), file.readAll());
    }
    if (type == VfsItemType::Image) {
        return listDirectory(info.absolutePath());
    }
    return {};
}

QByteArray Vfs::readBytes(const QString& uri) const
{
    ensureNestedDepthAllowed(uri, policy_);
    if (isArchiveUri(uri)) {
        const ArchiveUriParts parsed = parseArchiveUri(uri);
        if (isArchiveUri(parsed.archiveUri)) {
            const QByteArray nestedArchive = readBytes(parsed.archiveUri);
            return zipReader_.readMember(nestedArchive, parsed.memberPath);
        }
        QFile archive(pathFromFileUri(parsed.archiveUri));
        if (!archive.open(QIODevice::ReadOnly)) {
            throw std::runtime_error("Failed to open archive");
        }
        return zipReader_.readMember(archive.readAll(), parsed.memberPath);
    }

    QFile file(pathFromFileUri(uri));
    if (!file.open(QIODevice::ReadOnly)) {
        throw std::runtime_error("Failed to open file");
    }
    return file.readAll();
}

std::vector<VfsItem> Vfs::listDirectory(const QString& path) const
{
    QDir dir(path);
    const QFileInfoList entries = dir.entryInfoList(QDir::Files | QDir::Dirs | QDir::NoDotAndDotDot, QDir::DirsFirst | QDir::Name);
    std::vector<VfsItem> result;
    for (const QFileInfo& entry : entries) {
        VfsItemType type = entry.isDir() ? VfsItemType::Directory : classifyPath(entry.absoluteFilePath());
        if (type == VfsItemType::Other) {
            continue;
        }
        result.push_back({
            fileUriFromPath(entry.absoluteFilePath()),
            entry.fileName(),
            type,
            static_cast<std::uint64_t>(entry.size()),
            static_cast<std::uint64_t>(entry.size()),
            entry.lastModified().toMSecsSinceEpoch(),
        });
    }
    return result;
}

std::vector<VfsItem> Vfs::listArchiveUri(const QString& uri) const
{
    ensureNestedDepthAllowed(uri, policy_);
    const QByteArray archiveData = readBytes(uri);
    return zipReader_.list(uri, archiveData);
}

VfsItemType Vfs::classifyPath(const QString& path) const
{
    const QString lower = path.toLower();
    if (lower.endsWith(".zip") || lower.endsWith(".cbz")) {
        return VfsItemType::Archive;
    }
    static const QStringList imageExts = {
        ".png", ".jpg", ".jpeg", ".webp", ".bmp", ".gif", ".tif", ".tiff", ".avif"
    };
    for (const QString& ext : imageExts) {
        if (lower.endsWith(ext)) {
            return VfsItemType::Image;
        }
    }
    return VfsItemType::Other;
}
